proposals/0000-fine-grained-unused-warnings.rst

```diff
@@ -0,0 +1,394 @@
+       where bar2 = True
+
+   In this example
+
+   - ``foo2`` is directly unused
+   - ``foo1`` is indirectly unused, because it only occurs in the body of the unused ``foo2``, *and* ``foo1`` is in scope at the point of ``foo2``'s definition.
+   - Similarly, ``bar1`` is indirectly unused.
+   - But ``bar2`` is *not* indirectly unused, because, while it occurs in the body of the unused ``foo2``, ``bar2`` is not in scope at ``foo2``'s definition site.
+
+4. A **variable** *V* **bound by a pattern match**, assuming ``-Wunused-matches`` is enabled,
+
+   - is directly unused if it does not appear in the alternative the pattern match belongs to
+   - is indirectly unused if it referenced only in the body of a (directly or indirectly) unused bindings *C*, *and* *V* is in scope at the point where *C*'s definition appears.
+
+   For example, suppose ``bar1`` and ``bar2`` appear nowhere else.
+
+   ::
+
+     bar1 (Just v1) = undefined
+     bar2 (Just v2) v3 = v3
+       where c = v2
+
+   In this example
+
+   - ``c`` is directly unused
+   - ``v1`` is directly unused
+   - ``v2`` is indirectly unused
+   - ``v3`` is *not* indirectly unused, because, while it only occurs in the unused ``bar2``, ``v3`` is not in scope at the at ``bar2``'s definition site.
+
+5. An **imported identifier**, assuming ``-Wunused-imports`` is enabled,
+
+   - is directly unused if it is not mentioned anywhere in the module
+   - is indirectly unused if it is referenced only in (directly or indirectly) unused bindings
+
+6. A **forall-bound type variable**, assuming ``-Wunused-foralls`` is enabled,
+
+   - is directly unused if it does not appear in the body of the type
+   - is indirectly unused if it only appears in the kind signature of other (directly or indirectly) unused ``forall``-bound type variables in the body of the type
+
+   For example:
+
+   ::
+
+     far :: forall a (b :: a) c . c
+
+    Here, ``b`` is directly unused, but ``a`` is indirectly unused.
+
+**Warning References and Messages:**
+
+- A binding will produce a warning if
+
+  - it is directly unused, or
+
+  - it is indirectly unused and ``-Windirectly-unused-binds`` is enabled
+
+- The warnings for (directly or indirectly) unused bindings will reference all bindings they are used in. For example, if ``-Wunused-top-binds`` and ``-Wunused-local-binds`` are enabled,
+
+  ::
+
+    foo = bar
+    baz = pureStrLn "Hi"
+      where quux = bar
+    bar = ...
+    main = baz
+
+  will produce three warnings:
+
+  - ``foo`` is directly unused
+
+  - ``quux`` is directly unused
+
+  - ``bar`` is indirectly unused, and will produce a warning stating
+
+    ::
+
+      warning: [-Wunused-top-binds, -Windirectly-unused-binds]
+          ‘bar' is defined but used only in the following unused bindings: ‘foo’, ‘quux’
+
+- The warning for an unused binding B will reference the innermost (directly or indirectly) unused binding(s) whose right-hand sides mention B. For example, suppose ``bar`` is
```

Comment:
I think I understand what this rule wants to say, but I think it is a bit unclear. I think because this is an amendment/adding precision to the previous point without explicitly saying so.

Maybe this formulation reflects the meaning more precisely?
```suggestion
- If the warning for an unused binding B would reference multiple nested bindings it will only reference the innermost (directly or indirectly) unused binding(s) of those. For example, suppose ``bar`` is
```